TF version compatibility fixes

Rocketknight1 opened the pull request:
This PR makes several changes to core methods to support the upcoming 2.13 release of TF and generally futureproof against any other upcoming changes that might happen.

The core source of all of these problems is that Keras has been very mobile inside TensorFlow: Initially we used `tf.python.keras` until they told us this was a deprecated copy of Keras that shouldn't be there, and it got removed. We then switched to `tf.keras`, but now Keras has fully moved into its own library and namespace again. Although this is still mirrored at `tf.keras`, for the newest version of TF we'll just `import keras`.

There are several other related problems where our code assumed that parts of Keras that weren't really part of the public API would stay where they were. Not all of these have caused problems yet (that I know of) but they look very risky to me, and so I made some general fixes. This might surface some hidden bugs!

Fixes #23352

Rocketknight1 committed:
New TF version compatibility fixes

Rocketknight1 commented on src/transformers/modeling_tf_utils.py:
```diff
@@ -102,6 +101,18 @@
 ]
 
 
+def expand_1d(data):
```
This function was part of the Keras code that is not really in the public API. I just copied the whole thing rather than trying to import a moving target.

sgugger replied:
Should be in `tf_utils` though.

Rocketknight1 replied:
Moved!

Rocketknight1 commented on src/transformers/pipelines/base.py:
```diff
@@ -278,7 +278,7 @@ def infer_framework_load_model(
             raise ValueError(f"Could not load model {model} with any of the following classes: {class_tuple}.")
 
-    framework = "tf" if "keras.engine.training.Model" in str(inspect.getmro(model.__class__)) else "pt"
+    framework = "tf" if "TFPreTrainedModel" in str(inspect.getmro(model.__class__)) else "pt"
```
This code also assumed internal Keras namespaces wouldn't move around. I've used "TFPreTrainedModel" as a string that should be more constant, but we could also just use "tensorflow" or "keras" if there are cases where the models we want to use will not inherit from `TFPreTrainedModel`.

Rocketknight1 replied:
@sgugger this is one bit I was uncertain about - does every model we could use in a pipeline inherit from `PreTrainedModel` or `TFPreTrainedModel`? I assume yes but it felt a little dangerous.

Rocketknight1 replied:
Alright, let me try to figure out something that covers all the relevant versions without catching anything else.

Rocketknight1 commented on src/transformers/tf_utils.py:
```diff
@@ -166,3 +166,78 @@ def check_embeddings_within_bounds(tensor: tf.Tensor, embed_dim: int, tensor_nam
     )
+
+
+def save_attributes_to_hdf5_group(group, name, data):
```
These two functions are also moving targets I was trying to import from Keras - I've just given up and copied them into the transformers codebase.

sgugger commented on src/transformers/modeling_tf_utils.py:
```diff
@@ -507,6 +518,7 @@ def input_processing(func, config, **kwargs):
         if isinstance(v, allowed_types) or v is None:
             output[k] = v
         else:
+            print()
```
To remove

Rocketknight1 replied:
Removed, and moved expand_1d!

sgugger commented on src/transformers/modeling_tf_utils.py:
```diff
@@ -1528,8 +1536,8 @@ def train_step(self, data):
-            data = data_adapter.expand_1d(data)
-        x, y, sample_weight = data_adapter.unpack_x_y_sample_weight(data)
+            data = expand_1d(data)
+        x, y, sample_weight = tf.keras.utils.unpack_x_y_sample_weight(data)
```
Has this been here for all the TF versions we support?

Rocketknight1 replied:
Yes, since at least 2.4!

sgugger reviewed (commented):
Thanks for working on this!

Rocketknight1 committed:
Remove dummy print statement, move expand_1d

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_23663). All of your documentation changes will be reflected on that endpoint.

Rocketknight1 committed:
Make a proper framework inference function

Rocketknight1 committed:
Make a proper framework inference function

Rocketknight1 commented:
@sgugger I've added a proper framework inference function, put it in `utils/generic` and moved everything over to that. That should catch any edge cases in future, unless one of the frameworks gets renamed entirely :sweat_smile:

Rocketknight1 committed:
ValueError -> TypeError

sgugger reviewed (approved):
Thanks for iterating!
